### DISCUSSION TITLE
XState in libraries

### DISCUSSION THREAD

**Discussion by grimnomad** [statelyai/xstate#2648] [Q&A] [answered]
I have a question in regard of XState and its usage in libraries. I have a monorepo managed by Lerna. This monorepo contains a library and a server application. Both are useing XState. The library exports a function, which creates a state machine. If the state machines gets interpreted and executed in the server, it doesn't behaves as expected. The state machine spawns a callback actor, but actor does not execute as i would expect.

Here's a [minimal repsitory](https://github.com/grimnomad/xstate-lib-monorepo).

My question is, do I miss something or why does the callback actor doesn't transition?

**Comment by davidkpiano** [comment, marked as answer]
Have you gotten this to work without being tied to a framework? Extract the machine and test it in isolation.

**Reply by grimnomad** [reply]
Yes, if I locate the machine in the server, it executes as expected.

**Reply by grimnomad** [reply]
I think I misunterstood your question at first. I created a repository without any frameworks (except lerna). It shows the same behaviour. Here is the repository: [https://github.com/grimnomad/xstate-lib-test](https://github.com/grimnomad/xstate-lib-test)

**Reply by grimnomad** [reply]
Hey @davidkpiano, did you have time to look at this?

**Reply by davidkpiano** [reply]
Have you tried a top-level `invoke: { ... }` instead? That pattern seems to fit better here, and may solve your problem.

**Reply by grimnomad** [reply]
No, I haven't tried it yet. I will give it a try.

**Reply by grimnomad** [reply]
The `invoke: {...}` does solve the problem.
